lib/cn/node_split.c

```diff
@@ -366,3 +370,85 @@ cn_tree_node_get_split_key(
 
     return err;
 }
+
+static void
+kvset_split_res_init(struct cn_compaction_work *w, struct kvset_split_res *result, uint ks_idx)
+{
+    memset(result, 0, sizeof(*result));
+
+    for (int i = LEFT; i <= RIGHT; i++) {
+        uint idx = (i == LEFT ? ks_idx : ks_idx + w->cw_kvset_cnt);
+
+        result->ks[i].blks = &w->cw_outv[idx];
+        blk_list_init(&result->ks[i].blks->kblks);
+        blk_list_init(&result->ks[i].blks->vblks);
+        result->ks[i].blks->bl_vused = 0;
+
+        result->ks[i].blks_commit = &w->cw_split.commit[idx];
+        blk_list_init(result->ks[i].blks_commit);
+
+        result->ks[i].vgmap = &w->cw_vgmap[idx];
+    }
+
+    result->blks_purge = &w->cw_split.purge[ks_idx];
+    blk_list_init(result->blks_purge);
+}
+
+static void
+kvset_split_res_free(struct kvset *ks, struct kvset_split_res *result)
+{
+    for (int i = LEFT; i <= RIGHT; i++) {
+        blk_list_free(&result->ks[i].blks->kblks);
+        blk_list_free(&result->ks[i].blks->vblks);
+        blk_list_free(result->ks[i].blks_commit);
+    }
+
+    blk_list_free(result->blks_purge);
+}
+
+merr_t
+cn_split(struct cn_compaction_work *w)
+{
+    struct kvset_list_entry *le;
+    struct key_obj split_kobj;
+    struct cndb *cndb = cn_tree_get_cndb(w->cw_tree);
+    merr_t err;
+    uint i;
+
+    err = cn_tree_node_get_split_key(w->cw_node, w->cw_split.key, HSE_KVS_KEY_LEN_MAX,
+                                     &w->cw_split.klen);
+    if (err)
+        return err;
+
+    key2kobj(&split_kobj, w->cw_split.key, w->cw_split.klen);
+
+    for (i = 0, le = list_first_entry_or_null(&w->cw_node->tn_kvset_list, typeof(*le), le_link);
+         i < w->cw_kvset_cnt;
+         i++, le = list_next_entry(le, le_link)) {
+
+        struct kvset *ks = le->le_kvset;
+        struct kvset_split_res result = { 0 };
+
+        if (atomic_read(w->cw_cancel_request))
+            return merr(ESHUTDOWN);
```

Comment:
Why bother?  This should be fast, right?

Reply:
I did this following what other compaction functions do.. I don't know what happens otherwise, will check.

Reply:
I can move it up before the loop for now.

From an IO standpoint, the split operation would at most write 2 kblocks (1 kblock worth of data), 1 vblock, 1 hblock and a couple of cNDB records for each kvset. I presume that should be fast.  We can bring it back within the loop later if needed.

Reply:
I feel like there's no point if it's not in the loop.   So either leave it in the loop or get rid of it.  I see no harm in leaving it.

Reply:
Right.  But they prob have a much longer run-time than split.  But looking at cn_spill() I think it checks waaay too often, the check should be moved into the next outer block AFAICT...

Reply:
> I feel like there's no point if it's not in the loop.

hmm. Not sure if I follow that.. We can avoid starting a split when a close is in progress by checking it before the loop, right?

I see that 'cw_canceled' is being set for canceling a compaction operation. 'cw_canceled' is set only when:
1. rspills are wedged and
2. The core compaction functions (cn_split/cn_spill etc.) returns an ESHUTDOWN.

Reply:
Agreed.  If we expected the loop to iterate for a minute or longer (like spills often do) then we should check within the loop (as we do now, but in the innermost loop which unnecessary and wasted effort on an already overburdended task).
With split we expect to loop maybe 4-12 iterations, each iteration being very quick, so canceling in the middle of the split just makes a mess that we have to clean up later, much better IMO to let it finish.

Reply:
I assume the scheduler will not start operations once a kvs close has been initiated, which is why I said there's no point in checking it before you start the loop.  But I hadn't thought about the wedge scenario.

Reply:
I still don't understand rspill wedging, but I will probably need to in order to implement the split/spill synchronization...